Add methods to climate entity. (#2057)
home-assistant/developers.home-assistant: docs/core/entity/climate.md

        """Set new target hvac mode."""
```

> ### Turn on

Reviewer: We should explain that the expected outcome of a successful `climate.turn_on` call is that the hvac_mode is different from `HVACMode.OFF`, either optimistically set by the service handler or with the next state update.

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods if the device supports it.

>     # Make sure that hvac_mode is properly adjusted as well.

Reviewer: This comment, and the similar one for turn off is a bit unclear / misleading. The hvac_mode doesn't have to be optimistically set by the `turn_on` or `async_turn_on`, it should instead be set when we get a state update from the remote device in most cases.

Author: Yes, that is what I meant here. I just had trouble wording that clearly and yet succinctly so I wrote it in the most general terms I could find.
Maybe you have some proposals for better phrasing?

    def turn_on(self):
        """Turn the entity on."""

    async def async_turn_on(self):
        """Turn the entity on."""
```

### Turn off

> 

Reviewer: We should explain that the expected outcome of a successful `climate.turn_off` call is that the hvac_mode is `HVACMode.OFF`, either optimistically set by the service handler or with the next state update.

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods if the device supports it.

>     # Make sure that hvac_mode is set to Off.

Reviewer: If this comment is kept, use the correct name: HVACMode.OFF

```

### Turn on

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods.
    # The `turn_on` method should set `hvac_mode` to any other than
    # `HVACMode.OFF` by optimistically setting it from the service handler
    # or with the next state update

    def turn_on(self):
        """Turn the entity on."""

    async def async_turn_on(self):
        """Turn the entity on."""
```

### Turn off

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods.
    # The `turn_off` method should set `hvac_mode` to `HVACMode.OFF` by
    # optimistically setting it from the service handler or with the next state update

    def turn_off(self):
        """Turn the entity off."""

    async def async_turn_off(self):
        """Turn the entity off."""
```

### Toggle

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods.
    # The `toggle` method should set `hvac_mode` to the right `HVACMode` by
    # optimistically setting it from the service handler or with the next state update

    def toggle(self):
        """Toggle the entity."""

    async def async_toggle(self):
        """Toggle the entity."""

> ```

Reviewer: We should explain it's not mandatory to implement toggle, if not implemented the base entity's will call turn_on or turn_off according to the current HVAC mode.

        """Set new target hvac mode."""
```

### Turn on

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods.
    # The `turn_on` method should set `hvac_mode` to any other than
    # `HVACMode.OFF` by optimistically setting it from the service handler
    # or with the next state update

    def turn_on(self):
        """Turn the entity on."""

    async def async_turn_on(self):
        """Turn the entity on."""
```

### Turn off

```python
class MyClimateEntity(ClimateEntity):
    # Implement one of these methods.
    # The `turn_off` method should set `hvac_mode` to `HVACMode.OFF` by
    # optimistically setting it from the service handler or with the next state update

    def turn_off(self):
        """Turn the entity off."""

    async def async_turn_off(self):
        """Turn the entity off."""
```

### Toggle

```python
class MyClimateEntity(ClimateEntity):

>     # Implement one of these methods.
>     # The `toggle` method should set `hvac_mode` to the right `HVACMode` by
>     # optimistically setting it from the service handler or with the next state update.
> 
>     # It's not mandatory to implement the `toggle` method as the base implementation
>     # will call `turn_on`/`turn_off` according to the current HVAC mode.

Reviewer: ```suggestion
    # It's not mandatory to implement the `toggle` method as the base implementation
    # will call `turn_on`/`turn_off` according to the current HVAC mode.

    # If implemented, the `toggle` method should set `hvac_mode` to the right `HVACMode` by
    # optimistically setting it from the service handler or with the next state update.
```
